### PR TITLE
Make the project become a humilis layer

### PR DESCRIPTION
**Pull request by germangh**
This should allow us to easily include this repo into our AWS environment, at least once we have one properly defined :).

By the way, I would also suggest renaming the repo to `humilis-snapshots`, same as I have done with the `humilis-nat` repo that contains a layer (a CF stack) that provides support for a a new custom CF resource. The idea is to later have a parent "environment" repository where we just add the various layer repos as subtrees. Example of layers that will go there:
- snapshots
- nat: Support for a custom CF resource (NatGateway)
- vpc: Our networking layout once we have set it up correctly
- iam: Our IAM stuff such as users, groups, roles, etc.
- tapi: The event tracking API
- dapi?: Maybe in the future. Since you should be able to deploy whatever beanstalk app you have using CF I guess this is possible.
- etc.

So each of those layers would have their own repository (`humilis-[layer_name]`) and it should be possible to update them relatively independently, as long as keep the interfaces between layers fixed. For instance we could agree on never using `layer` references (see the humilis docs), but using `output` references instead.

@bubenkoff what do you think about this way of managing our AWS resources?

[Review on Reviewable](https://reviewable.io/reviews/innovativetravel/auto-snapshot/6)

### CODE REVIEW

**Comment by bubenkoff** [comment]
Review status: 0 of 23 files reviewed at latest revision, 6 unresolved discussions.

---

*[lambda/.coverage, line 1 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvSv7O-6FGTTuKcfV:-K8YvSv8hgCbgOTzSYSa:861705787) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/lambda/.coverage#L1)):*
remove and add to ignores

---

*[lambda/Makefile, line 11 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvZEMU-Q_7KwC3xGf:-K8YvZEMU-Q_7KwC3xGg:-2106634075) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/lambda/Makefile#L11)):*
use tox and pytest

---

*[lambda/tests/__init__.py, line 0 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YviAcd2TdUxicp7Ux:-K8YviAcd2TdUxicp7Uy:1228961144) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/4cacd37a9d1623eac23e13ac4715af080fd71500/lambda/tests/__init__.py#L0)):*
move tests out of the actual code, should never be deployed

---

*[layers/autosnapshot, line 1 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvraFNqBdRilmXCRM:-K8YvraFNqBdRilmXCRN:166874) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/layers/autosnapshot#L1)):*
?

---

*[Makefile, line 7 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvyjGWEkDUnTmwHRO:-K8YvyjGWEkDUnTmwHRP:1750326127) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/Makefile#L7)):*
document targets with one-line commets

---

*[README.md, line 111 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YxT5MDGPgTK3YGy-s:-K8YxT5MDGPgTK3YGy-t:1489531884) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/README.md#L111)):*
how can we solve pluggability other than with a submodules?
we could use python entry points, or more sophisticated https://github.com/hpk42/pluggy

so basically we can turn this package into a proper python package with resources (data files), and it will plug into humillis

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6)*

**Comment by samuelazran** [comment]
*[README.md, line 6 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8Z4BLehrJ5-M_fbe4l:-K8Z4BLehrJ5-M_fbe4m:-1599536531) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/README.md#L6)):*
The "scheduled" part is optional and not part of the deployment (should it?). so still there is a manual work to set the event that trigger the lambda (cron job for instance)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6)*

**Comment by germangh** [comment]
Review status: 0 of 22 files reviewed at latest revision, 7 unresolved discussions.

---

*[lambda/Makefile, line 11 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvZEMU-Q_7KwC3xGf:-K8Z51ZxhVeK9MQW0M6p:1553492269) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/lambda/Makefile#L11)):*
Why using tox? Lambdas can only run in Python 2.7?

---

*[lambda/tests/__init__.py, line 0 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YviAcd2TdUxicp7Ux:-K8Z1QF7YzwlZR5VtwPB:-1579558243) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/4cacd37a9d1623eac23e13ac4715af080fd71500/lambda/tests/__init__.py#L0)):*
humilis removes any directory called tests before deploying  to AWS. So probably better to leave them where they are now?

---

*[layers/autosnapshot, line 1 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvraFNqBdRilmXCRM:-K8Z22OdUceP88Zeywep:-461229918) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/layers/autosnapshot#L1)):*
The reason for this is that if you run:
`humilis create humilis-autosnapshot.yaml`

then humilis will look for the `autosnapshot` layer in directory `layers/autosnapshot`. So I use a symlink to just make humilis look for the `autosnapshot` layer files in the root directory of the repository. 

---

*[Makefile, line 7 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvyjGWEkDUnTmwHRO:-K8Z2pdP4AIMeKlt5T9k:-499146009) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/Makefile#L7)):*
Done.

---

*[README.md, line 111 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YxT5MDGPgTK3YGy-s:-K8Z3WdQgZJXbObIaREs:1349909899) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/README.md#L111)):*
I was planning to use subtrees instead. So we would have a parent repository with an environment definition file, and the various layers in `layers/[layer_name]` where `layers/[layer_name]` contains repository `humilis-[layer_name]` as a subtree. I am not convinced by the option of having instead this and other similar `humilis-[layer_name]` plugging into humilis. It would be something like having ansible cookbooks plugging into ansible.

---

*[lambda/.coverage, line 1 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvSv7O-6FGTTuKcfV:-K8Z2rt-wMupcaVoJDdI:-499146009) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/lambda/.coverage#L1)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6)*

**Comment by germangh** [comment]
Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions.

---

*[README.md, line 6 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8Z4BLehrJ5-M_fbe4l:-K8Z5ZVSC-1s6BsIr-Dj:1406629128) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/README.md#L6)):*
IMO it should be part of the deployment: The cron schedule should be a layer parameter in `meta.yaml`. Without a schedule the auto-snapshot is just useless and you need to remember (and document) that you must manually go to AWS Lambda to set the schedule. We could easily have dozens of layers doing devops kind of things in the future and you don't want to remember which one you need to manually tune after deployment.

But adding the schedule to the layer definition is too much work for the payoff atm. So let's leave it for now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6)*

**Comment by bubenkoff** [comment]
Review status: 0 of 22 files reviewed at latest revision, 4 unresolved discussions.

---

*[lambda/Makefile, line 11 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvZEMU-Q_7KwC3xGf:-K8Z5WcCea_sJpd7r_Fs:-1091943874) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/lambda/Makefile#L11)):*
it is sort of de facto standard as an entry point for testing, but probably it's ok like this if you meant test target only for development needs and not for CI

---

*[lambda/tests/__init__.py, line 0 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YviAcd2TdUxicp7Ux:-K8Z6-1BdpMvwXrtL5kM:1672255954) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/4cacd37a9d1623eac23e13ac4715af080fd71500/lambda/tests/__init__.py#L0)):*
hmm, that's a surprising behaviour :) I would not do any removals for humillis

---

*[layers/autosnapshot, line 1 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvraFNqBdRilmXCRM:-K8Z6GYLOFjjwUk19leY:-1743735691) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/layers/autosnapshot#L1)):*
then you can add a relative symlink, not an absolute one

---

*[README.md, line 111 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YxT5MDGPgTK3YGy-s:-K8Z6kc_0xaISLpRcVhf:-1578471137) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/README.md#L111)):*
ok we can discuss this later, i guess

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6)*

**Comment by samuelazran** [comment]
*[layers/autosnapshot, line 1 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvraFNqBdRilmXCRM:-K8Z7SP_6RHWkpqa4LFh:1659403345) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/layers/autosnapshot#L1)):*
in the make file maybe

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6)*

**Comment by samuelazran** [comment]
Reviewed 26 of 29 files at r1, 2 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6)*

**Comment by samuelazran** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6)*

**Comment by germangh** [comment]
Review status: 15 of 18 files reviewed at latest revision, 3 unresolved discussions.

---

*[layers/autosnapshot, line 1 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvraFNqBdRilmXCRM:-K8ZZQlLNrwMKIPTI8hR:-499146009) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/layers/autosnapshot#L1)):*
Done.

---

*[lambda/Makefile, line 11 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YvZEMU-Q_7KwC3xGf:-K9chrFkouwP8vVYBZgl:-1887290062) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/fc9f145416a446c105e4dff04895cabc15475ac7/lambda/Makefile#L11)):*
Using tox again.

---

*[lambda/tests/__init__.py, line 0 \[r1\]](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6#-K8YviAcd2TdUxicp7Ux:-K8ZZSVVF0h1ETg2jNBp:-1527643163) ([raw file](https://github.com/innovativetravel/auto-snapshot/blob/4cacd37a9d1623eac23e13ac4715af080fd71500/lambda/tests/__init__.py#L0)):*
Not surprising to me :). But yes is probably a misfeature. Will fix that in the next humilis release. 

I moved the tests dir out of the lambda deployment package as you suggested.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/innovativetravel/auto-snapshot/6)*

**Comment by germangh** [comment]
I think this is ready to be merged, but please do a quick review if you have time.

**Comment by germangh** [comment]
Deprecating this repo in favor of AWS EBS scheduler:

https://aws.amazon.com/answers/infrastructure-management/ebs-snapshot-scheduler/
